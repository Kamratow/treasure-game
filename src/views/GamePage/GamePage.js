import React from "react";
import PropTypes from "prop-types";

import { connect } from "react-redux";

import GameBoard from "../../components/GameBoard/GameBoard";
import PageHeader from "../../components/PageHeader/PageHeader";
import TurnControl from "../../components/TurnControl/TurnControl";
import AddPlayer from "../../components/AddPlayer/AddPlayer";
import LeaderBoard from "../../components/LeaderBoard/LeaderBoard";

import * as actions from "../../store/actions/index";

import classes from "./GamePage.module.css";

export const GamePage = ({
  gameName,
  playerName,
  isGameStarted,
  currentTurn,
  movesLeft,
  isGameSetupLoading,
  gameSetup,
  isGameSetupFetchFailed,
  treasuresLeft,
  isGameFinished,
  scores,
  onPlayerNameChange,
  onGameStart,
  onCellClick,
  onTreasureClick,
  onRevealCells,
  onEndGame,
  onRestartGame
}) => {
  const nameChanged = newPlayerName => {
    onPlayerNameChange(newPlayerName);
  };

  const gameStarted = () => {
    onGameStart();
  };

  const cellClicked = (cell, index) => {
    if (cell.proximity === "T" && treasuresLeft === 1) {
      onEndGame();
      return;
    }

    if (cell.proximity === "T") {
      onTreasureClick();
    }

    if (movesLeft === 1) {
      onRevealCells(cell, index);
    } else {
      onCellClick(cell, index);
    }
  };

  let gamePageContent = null;

  if (isGameFinished) {
    gamePageContent = (
      <>
        <h3 className={classes.GamePageFinishTitle}>
          Congratulations {playerName}!! You have won in {currentTurn} turns.
        </h3>
        <button type="button" onClick={() => onRestartGame()}>
          Restart Game
        </button>
        <LeaderBoard scores={scores} />
      </>
    );
  } else {
    gamePageContent = (
      <>
        <PageHeader headerText={gameName} />
        <AddPlayer
          isGameStarted={isGameStarted}
          playerName={playerName}
          gameStarted={gameStarted}
          nameChanged={newPlayer => nameChanged(newPlayer)}
        />
        <TurnControl
          isGameStarted={isGameStarted}
          currentTurn={currentTurn}
          movesLeft={movesLeft}
        />
        <GameBoard
          isGameStarted={isGameStarted}
          isGameSetupLoading={isGameSetupLoading}
          isGameSetupFetchFailed={isGameSetupFetchFailed}
          gameSetup={gameSetup}
          cellClicked={cellClicked}
        />
      </>
    );
  }

  return gamePageContent;
};

const mapStateToProps = state => ({
  gameName: state.game.gameName,
  playerName: state.game.playerName,
  isGameStarted: state.game.isGameStarted,
  currentTurn: state.game.currentTurn,
  movesLeft: state.game.movesLeft,
  isGameSetupLoading: state.game.isGameSetupLoading,
  gameSetup: state.game.gameSetup,
  isGameSetupFetchFailed: state.game.isGameSetupFetchFailed,
  treasuresLeft: state.game.treasuresLeft,
  isGameFinished: state.game.isGameFinished,
  scores: state.game.scores
});

const mapDispatchToProps = dispatch => {
  return {
    onPlayerNameChange: playerName =>
      dispatch(actions.gameChangePlayerName(playerName)),
    onGameStart: () => dispatch(actions.gameStart()),
    onCellClick: (cell, cellIndex) =>
      dispatch(actions.gameCellSelected(cell, cellIndex)),
    onTreasureClick: () => dispatch(actions.gameTreasureClicked()),
    onEndGame: () => dispatch(actions.gameEnd()),
    onRevealCells: (cell, cellIndex) =>
      dispatch(actions.gameRevealCells(cell, cellIndex)),
    onRestartGame: () => dispatch(actions.gameRestart())
  };
};

GamePage.propTypes = {
  gameName: PropTypes.string,
  playerName: PropTypes.string,
  isGameStarted: PropTypes.bool,
  currentTurn: PropTypes.number,
  movesLeft: PropTypes.number,
  gameSetup: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.number,
      proximity: PropTypes.string
    })
  ),
  isGameSetupFetchFailed: PropTypes.bool,
  isGameFinished: PropTypes.bool,
  scores: PropTypes.arrayOf(
    PropTypes.shape({
      playerName: PropTypes.string,
      score: PropTypes.number
    })
  ),
  onPlayerNameChange: PropTypes.func,
  onGameStart: PropTypes.func,
  onGameFetchInitialSetup: PropTypes.func,
  onCellClick: PropTypes.func,
  onTreasureClick: PropTypes.func,
  onRevealCells: PropTypes.func,
  onEndGame: PropTypes.func,
  onRestartGame: PropTypes.func
};

export default connect(mapStateToProps, mapDispatchToProps)(GamePage);
