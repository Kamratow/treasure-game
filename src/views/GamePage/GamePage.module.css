.GamePageFinishTitle {
  margin-top: 50px;
}
