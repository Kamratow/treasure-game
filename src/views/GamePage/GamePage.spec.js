import React from "react";
import { shallow } from "enzyme";

import { GamePage } from "./GamePage";

const defaultProps = {
  gameName: "Test game",
  playerName: "Test player",
  isGameStarted: true,
  currentTurn: 1,
  movesLeft: 3,
  isGameSetupLoading: false,
  gameSetup: [
    {
      id: 1,
      proximity: "T"
    },
    {
      id: 2,
      proximity: "3"
    }
  ],
  isGameSetupFetchFailed: false,
  treasuresLeft: 3,
  isGameFinished: false,
  scores: [],
  onPlayerNameChange: jest.fn(),
  onGameStart: jest.fn(),
  onCellClick: jest.fn(),
  onTreasureClick: jest.fn(),
  onRevealCells: jest.fn(),
  onEndGame: jest.fn(),
  onRestartGame: jest.fn()
};

describe("GamePage", () => {
  it("renders properly when game is not finished", () => {
    const wrapper = shallow(<GamePage {...defaultProps} />);
    expect(wrapper.debug()).toMatchSnapshot();
  });

  it("renders properly when game is finished", () => {
    const propsToRender = {
      ...defaultProps,
      isGameFinished: true,
      scores: [
        { playerName: "Player1", score: 3 },
        { playerName: "Player2", score: 5 }
      ]
    };
    const wrapper = shallow(<GamePage {...propsToRender} />);
    expect(wrapper.debug()).toMatchSnapshot();
  });
});
