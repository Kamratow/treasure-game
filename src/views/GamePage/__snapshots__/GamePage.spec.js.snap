// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GamePage renders properly when game is finished 1`] = `
"<Fragment>
  <h3 className=\\"GamePageFinishTitle\\">
    Congratulations 
    Test player
    !! You have won in 
    1
     turns.
  </h3>
  <button type=\\"button\\" onClick={[Function: onClick]}>
    Restart Game
  </button>
  <LeaderBoard scores={{...}} />
</Fragment>"
`;

exports[`GamePage renders properly when game is not finished 1`] = `
"<Fragment>
  <PageHeader headerText=\\"Test game\\" />
  <AddPlayer isGameStarted={true} playerName=\\"Test player\\" gameStarted={[Function: gameStarted]} nameChanged={[Function: nameChanged]} />
  <TurnControl isGameStarted={true} currentTurn={1} movesLeft={3} />
  <GameBoard isGameStarted={true} isGameSetupLoading={false} isGameSetupFetchFailed={false} gameSetup={{...}} cellClicked={[Function: cellClicked]} />
</Fragment>"
`;
